assets/js/types/type-defs/product-response.ts

```diff
@@ -17,7 +17,8 @@ export interface ProductResponseItemPrices extends CurrencyResponse {
 }
 
 export interface ProductResponseItemData {
-	name: string;
+	name?: string;
+	key?: string;
```

Comment:
I'm wondering here… 🤔  this changes a bit the intent of the type. Previously, `name` was required. Now you could in theory have an object without either `name` or `key` and respect this interface; or, other way around, both could be defined.

Given the code above, it doesn't seem like a big deal as we treat the case in which both are `undefined` by making it into an empty string (and, also, we pass an empty string by default to the component anyways 😂 ), and we are giving priority to `key`; but I can't help but wonder if the intention of this interface should instead be that one of those two should be required, but not both or neither at the same time.

What are your thoughts?

Reply:
I agree. Whatever we're choosing as the correct property should be required. This will highlight issues in the codebase using the wrong property and allow future deprecation of the wrong one.

Reply:
Or, if both are supposed to be valid, but only one at the time, that should be also reflected in the type, IMHO.

Reply:
Thanks for the comments. I agree with you and I did initially investigate what the solution would be to make one of `key` or `name` required, but in my opinion it seems a bit convoluted, adding more complexity than necessary - see solution here:

https://stackoverflow.com/questions/40510611/typescript-interface-require-one-of-two-properties-to-exist

I'm fairly new to typescript so I can't think of a better way to do this but I'm open to suggestions. As an alternative, I think what Mike suggested makes sense of requiring `key` prop and having `name` as optional and marked as deprecated? What do you think @mikejolley @sunyatasattva ?

```
	key: string;
	name?: string; // deprecated

Reply:
Hey @alexflorisca, I agree TypeScript might feel convoluted at times, but in my experience it often pays off. I have experimented a bit and came up with a generic type that's hopefully a bit less convoluted. However, if this is still too much, I provide a non-generic solution below that will solve your problem while being fairly simple to read:

```
type EitherProperty<
    T, A extends keyof T, B extends keyof T
> = (Omit<T, B> & { [K in A]-?: T[A]; } & { [K in B]?: never} )
| (Omit<T, A> & { [K in B]-?: T[B] } & { [K in A]?: never });
```

See [<kbd>TypeScript Playground</kbd>](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgApQPYBMCuCwBKEAzgA4YjEQCSkAtgEJxUAicYcyA3gFDL-IA1hACeAfgBcyYmCigA5gG4+AkHDoRJ02QuUBIAG5wANjghSZckEp56swMsbjiLO6-oAWwLFgggtAEYYGMYQcCDKAL48PGAipCgAosBgHtDoGAlQcQA8KvwAKgA0yACCyBAAHpAgWMRCohgwyMXIDBXVfnUNIk0tPAB8yAC8yAAUAPJ0KTmtDEMAZNzIANoA0sigZQC6ALRaBSul24rIkchLXKsbWwzbWiAQBtDnAJQ8AD7jUzOtpYvLdabEBtPYHFZ3M4XQE3EHHB5PaBnV7KHgICgyTb0ADqwGMxkSUEwUCkyVS6UwWVyGVw+CIZAxNHoTFY7DgJQARMIRBzOWoNByhqNeAIelIOTBgrz8sh+eZkByAnAoNLDCYzOKjMZVfZHM5xbrSKqvD4-FJ4MYqEUeNE0RiwFiIHRcfiAHIYMCE4lrUSklJpKAZKkiHI0vCEEjkShMp0siBsDic7m8hVywUjbgy7niyUYVVajUKrU6hykJwiA2l43eXwgc0mK02mLoygOlJOl3Gd2eokYKCu9TyskBoPQamYWkRhnR2ix5jxtlJ0QpjlpoWZ0VynNS61q0zyjnF3eG8uVsjV0115AWxvRIA) on how to use this.

Alternatively, if this feels too complex, what about this?

```
interface ProductResponseItemBaseData {
	value: string;
	display?: string;
	hidden?: boolean;
}

type ProductResponseItem = ProductResponseItemBaseData & ({ key: string, name?: never } | { key?: never, name: string });
```

See [<kbd>TypeScript Playground</kbd>](https://www.typescriptlang.org/play?ssl=8&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgApQPYBMCuCwBKEAzgA4YjEQCSkAtgEJxUAicYcyA3gFACQANzgAbHBABcyYmCigA5gG5+WYGWFwAngH5J02SEX8AFsCxYIIHcgBGGDMIhwQSgL48eYDaRTpseQiTklDT0yAC8aJi4+ERkFFS0EIzMEGwcyABkyAAUXMgA1hAaujLyADTIIHB0EFYgEALQyC7IAD7cBUV1DdAVVTUl+nLNAJRKPAjxYMjA9ADqwMLCAKJQmFCSvtEBccGJdOHcPMgnncXIAEQwdhdlx6f9EpfEt-cnQqJPFx+vp8gqak0kguWF+pxMZgskngwioPDcEymM3mi2EADkMGBVusANJFTZRfyxIIJUIRXh-QrnK43O5-D5iYE-OmnAGkdTU0Esk4Q8wgaEiOEIyaUaazJILJYYrFrDBQNHVJ5bImBeIhJKHCkPRXAl7c5AMr7Mt7-VTsoGXLkm3lQ5AwoU8IA) on how to use this.

Basically, how this works is that you are defining your basic data in your interface and completely excluding the name/key keys, and then you create the full object type by telling TS it is either one shape or the other.

Reply:
Thanks for this suggestions @sunyatasattva, this somehow fell off my radar so sorry for taking so long to update. I've updated the PR with your second suggestion of defining a base interface and a type that makes one of `name` or `key` mandatory. I think it reads nicely and is friendly to non typescript experts such as myself :). Appreciate your help on this :)